code/src/ng_link/ng_layer.py

```diff
@@ -0,0 +1,373 @@
+        
+        # Add all necessary ui controls here
+        ui_controls = [
+            f"#uicontrol {vec} color color(default=\"{color}\")",
+            "#uicontrol invlerp normalized",
+        ]
+        
+        # color emitter
+        emit_color = "void main() {\n" + f"emit{emitter}(color * normalized());" + "\n}"
+        shader_string = ""
+        
+        for ui_control in ui_controls:
+            shader_string += ui_control + '\n'
+        
+        shader_string += emit_color
+        
+        return shader_string
+        
+    def __set_shader(self, shader_config:str) -> None:
+        """
+        Sets a configuration for the neuroglancer shader.
+        
+        Parameters
+        ------------------------
+        shader_config: str
+            Shader configuration for neuroglancer in string format.
+        
+        Raises
+        ------------------------
+        ValueError:
+            If the provided shader_config is not a string.
+        
+        """
+        
+        # #uicontrol vec3 color color(default=\"green\")\n#uicontrol invlerp normalized\nvoid main() {\n  emitRGB(color * normalized());\n}
+        if not utils.check_type_helper(shader_config, str):
+            raise ValueError(f"Shader accepts only str. Received value: {value}")
+        
+        self.layer_state['shader'] = shader_config
+        
+    def __set_shader_control(self, shader_control_config:dict) -> None:
+        """
+        Sets a configuration for the neuroglancer shader control.
+        
+        Parameters
+        ------------------------
+        shader_control_config: dict
+            Shader control configuration for neuroglancer.
+        
+        Raises
+        ------------------------
+        ValueError:
+            If the provided shader_control_config is not a dictionary.
+        
+        """
+        
+        if not utils.check_type_helper(shader_control_config, dict):
+            raise ValueError(f"Shader control accepts only dict. Received value: {value}")
+        
+        self.layer_state['shaderControls'] = shader_control_config 
+    
+    def __set_image_channel(self, channel:int) -> None:
+        """
+        Sets the image channel in case the file contains multiple channels.
+        
+        Parameters
+        ------------------------
+        channel: int
+            Channel position. It will be incremented in 1 since neuroglancer channels starts in 1.
+        
+        Raises
+        ------------------------
+        ValueError:
+            If the provided channel is not an integer.
+        
+        """
+        
+        if not utils.check_type_helper(channel, int):
+            raise ValueError(f"Channel accepts only integer. Received value: {value}")
+
+        self.layer_state['localDimensions'] = {
+            "c'": [
+                channel + 1,
+                ""
+            ]
+        }
+        
+    def __set_visible(self, visible:bool) -> None:

```

Comment:
this looks like the `@property` pattern.  Do you want to do something like this instead?

```
@visible.setter
def visible(self, visible:bool):
  self.layer_state['visible'] = visible
```

Reply:
you could do the same for the other setters/getters (e.g. `__set_image_channel`)